### PR TITLE
lib/mempool: Fix spurious -ENOMEM due to agressive latency control

### LINKED ISSUE #11022: mempool allocator is broken
There is a bug in mempool implementation. When memory is allocated allocator gets an available block and if it is large enough it splits these block into 4 smaller blocks. This is repeated until allocator reaches the smallest block that satisfies the request. When `block_free` is called situation is reversed and when 4 neighboring blocks are found to be free allocator concatenates them and returns to the pool as one larger block.

The problem is that to avoid the locking allocator allows a situation when big block or 4 smaller blocks are taken from the pool before the split or merge. Author seemed to know about this limitation and added `-EAGAIN` that would allow alloc to retry but:
- first, when there are no free blocks `-ENOMEM` is returned: even though these blocks may be at the moment be reserved for concatenation `_sys_mem_pool_block_alloc` will not be able to see them, the same happens when `block_free` is merging blocks,
- second, this solution is non-stable: when free is performed by lower priority context and will be interrupted by high priority context just before `block_free` is about to be recursively called, memory will never be reported as free even if we switch to `-EAGAIN` in the `_sys_mem_pool_block_alloc`.

To replicate the problem with higher probability create two context competing for allocation (threads or thread/interrupt) and add busy wait just before `block_free` is recursively called.

There is also question of performance impact. Note that at the moment code is merging smaller blocks in every call to free and break it again with every alloc. Note that in normal situation application will be allocating small blocks only to free them short time later. Wouldn't it be better to merge blocks only when application requests bigger chunk?

I also think that operation of split and merge should be done atomically with free/alloc. If merge is done only on big chunk request the impact may not be big and the code would be stable.

### PR DESCRIPTION
**Pull request by andyross**
The mempool operations need to be atomic, but because of latency
concerns (the allocator is intended for use in an ISR) the locking was
designed to be as minimal as possible.  And it... mostly got it right.
All the list handling was correctly synchronized.  The merging of four
child blocks into a parent block was atomic.  The splitting of a block
into four children was atomic.

BUT: there was a moment between the allocation of a large block and
the re-addition of its three unused children where the lock was being
released.  This meant that another context (e.g. an ISR that just
fired, interrupting the existing call to k_mem_pool_alloc()) would see
some memory "missing" that wasn't actually allocated.  And if this
happens to have been the top level block, it's entirely possible that
the whole heap looks empty, even though the other allocator might have
been doing only the smallest allocation!

Fix that by making the "remove a block then add back the three
children we don't use" into an atomic step.  We can still relax the
lock between levels as we split the subblocks further.

(Finally, note that this trick allows a somewhat cleaner API as we can
do our "retry due to race" step internally by walking back up the
block size list instead of forcing our caller to do it via that weird
-EAGAIN return value.)

Fixes #11022

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

### CODE REVIEW

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

**Review by tyrantlizard77 — APPROVED** [review]

File: lib/mempool/mempool.c [hunk 1]
```diff
@@ -255,26 +249,36 @@ int _sys_mem_pool_block_alloc(struct sys_mem_pool_base *p, size_t size,
...
+				data = block_break(p, data, from_l, lsizes);
+				pool_irq_unlock(p, key);
+				key = pool_irq_lock(p);
+			}
+			break;
+		}
 	}
+	pool_irq_unlock(p, key);
 
 	*level_p = alloc_l;
```
Comment (ceolin) [inline lib/mempool/mempool.c]: level_p is unsigned but alloc_l is signed and if there is no available block its value will be -1.

Comment (ceolin) [reply lib/mempool/mempool.c]: ok, just saw that this is not possible, though, you're still assigning an unsigned to signed. Either you cast it or change alloc_l to be signed (requiring some logic change).

**Review by ceolin — APPROVED** [review]
The patch itself doesn't introduced the problem I've mentioned. So +1

**Review by pizi-nordic — APPROVED** [review]

**Review by pdunaj — CHANGES_REQUESTED** [review]
It seems race with two allocations will be fixed. There is still a case with `block_free` vs `_sys_mem_pool_block_alloc` race.
If interrupt will come while thread is in line 180 whole memory can disappear and allocation will fail.

**Comment by carlescufi** [comment]
@andyross can you take a look at [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/11220#pullrequestreview-173375489)?

**Comment by andyross** [comment]
I see it and agree.   The "remove three sibling blocks and add back the combined one" needs to be atomic and it isn't currently (there is a gap between the removal and addition, so an unlocky allocator at that moment will see missing memory).  Needs a little rework to how the merge loop is written, because it defeats the existing tail recursion trick I was so proud of.

**Comment by carlescufi** [comment]
@andyross thanks for the fix so far. Any idea when you will be able to take a look at the additional failure? Folks at Nordic are hitting this bug. Thanks!

**Comment by andyross** [comment]
Rebased.  Added an extra patch for the extra case @pdunaj caught.

**Comment by pdunaj** [comment]
@andyross , thanks! This look ok now and should work. Let me quickly test it...

**Comment by pdunaj** [comment]
I have checked the patch and it works fine :)
Thanks again @andyross .

**Review by pdunaj — APPROVED** [review]
I suggest squashing two commits into one.

**Comment by carlescufi** [comment]
@pdunaj I actually think the 2 commits make sense, since they fix 2 different issues. I will merge as-is.

**Pull request merged**